openmicroscopy/bioformats
# Codegen - Adding BinData

This PR will allow for our codegen to handle raw binary data. For the Mask implementation in particular this requires some unique behaviour to be added. In our schema a Mask extends base64binary, which in Java would be the equivalent of extending a byte[]. The metadata API reflects this by allowing you to set BinData as such - setMaskBinData(byte[] binData, int ROIIndex, int shapeIndex)

From a Java perspective it would make more sense to have the raw binary as an additional attribute on the BinData object, however this would require larger schema changes and break the existing API in favour of adding new functions that perhaps don't accurately reflect the schema. 

As such I have modified the codegen so that if a model object has a primitive base then it shall instead extend the default type and the primitive base will be added as an attribute with the tag baseAttribute. Properties tagged as being baseAttributes will be exposed in the API at the top level, ie setMaskBinData as opposed to setMaskBinDataBinData. 

This will see the BinData model object add the following property along with getters and setters. The getters and setters will then be used by the top level API functions, leaving the metadata API unchanged.
// BinData property
private byte[] binData;

Reasons for taking this approach:
- Limit schema changes as much as possible
- Metadata API should remain unchanged
- Simplifies templates by removing previous specialisations for shape and lightsource abstract 
- Removes last remaining hardcoded bin data functions

In Java we are using a raw byte[], the reason for using this as opposed to List<Byte>, ByteBuffer, InputStream etc was firstly for efficiency in storage and in performance when converting back and forth from the XML. In CPP this is currently listed as a string for simplicity but this would certainly welcome suggestions on a suitable alternative.

Note: The CPP implementation requires the Shape complex type PR - https://github.com/openmicroscopy/bioformats/pull/2160 
As such the CPP is not going to validate against the 2013 schema

Testing:
I am in the process of defining a suitable testing strategy.
- Check that all tests remain green
- Verify that the new attribute has been added to BinData model along with getters and setters
- Verify that metadata API remains unchanged
